[REVIEW] OO interface for hash join with explicit build/probe semantic

zanmato1984 opened the pull request:
As of issue #1087 requests, a more flexible hash join interface that exposes explicit build/probe semantics is desired.

This change is mainly inspired by the proposal in https://github.com/rapidsai/cudf/issues/1087#issuecomment-646093018, including:
1. An OO interface for hash join that builds the hash table in constructor and probes in individual `*_join()` member functions;
2. Decomposition of the old function chain from `join_call_compute_df` to `get_base_hash_join_indices` into individual build/probe functions;
3. OO implementation of hash join using the new build/probe functions above;
4. Old hash-based inner/left/full join APIs re-implemented using the OO hash join interface, making them aliases of the OO hash join interface;
5. Join UT updated to additionally test the build-once and probe-multiple (in parallel) semantic.

However I didn't implement the following as mentioned in https://github.com/rapidsai/cudf/issues/1087#issuecomment-646093018:

> This would also enable @magnatelee's request to have an API to compute the output size of a join operation

This may need some more modification (possibly minor), and it could be done in the near future I suppose.

Please help to review. Thanks a lot!

Closes #1087 

<!--

Thank you for contributing to cuDF :)

Here are some guidelines to help the review process go smoothly.

1. Please write a description in this text box of the changes that are being
   made.

2. Please ensure that you have written units tests for the changes made/features
   added.

3. There are CI checks in place to enforce that committed code follows our style
   and syntax standards. Please see our contribution guide in `CONTRIBUTING.MD`
   in the project root for more information about the checks we perform and how
   you can run them locally.

4. If you are closing an issue please use one of the automatic closing words as
   noted here: https://help.github.com/articles/closing-issues-using-keywords/

5. If your pull request is not ready for review but you want to make use of the
   continuous integration testing facilities please label it with `[WIP]`.

6. If your pull request is ready to be reviewed without requiring additional
   work on top of it, then remove the `[WIP]` label (if present) and replace
   it with `[REVIEW]`. If assistance is required to complete the functionality,
   for example when the C/C++ code of a feature is complete but Python bindings
   are still required, then add the label `[HELP-REQ]` so that others can triage
   and assist. The additional changes then can be implemented on top of the
   same PR. If the assistance is done by members of the rapidsAI team, then no
   additional actions are required by the creator of the original PR for this,
   otherwise the original author of the PR needs to give permission to the
   person(s) assisting to commit to their personal fork of the project. If that
   doesn't happen then a new PR based on the code of the original PR can be
   opened by the person assisting, which then will be the PR that will be
   merged.

7. Once all work has been done and review has taken place please do not add
   features or make changes out of the scope of those requested by the reviewer
   (doing this just add delays as already reviewed code ends up having to be
   re-reviewed/it is hard to tell what is new etc!). Further, please do not
   rebase your branch on master/force push/rewrite history, doing any of these
   causes the context of any comments made by reviewers to be lost. If
   conflicts occur against master they should be resolved by merging master
   into the branch used for making the pull request.

Many thanks in advance for your cooperation!

-->

GPUtester commented:
Can one of the admins verify this patch?

jrhemstad commented:
Thanks for contributing @zanmato1984! We'll get on reviewing this.

zanmato1984 commented:
I'm going to resolve the conflicts with the base branch. Hold on reviewing.

zanmato1984 commented:
Conflicts resolved.

jrhemstad commented on cpp/include/cudf/join.hpp:
```diff
@@ -367,5 +367,131 @@ std::unique_ptr<cudf::table> cross_join(
+   * @param build_on The column indices from `build` to join on.
+   * @return Hash join instance.
+   */
+  static std::unique_ptr<const hash_join> create(cudf::table_view const& build,
```
Why is this a static factory method? I think this should just be the constructor for the `hash_join` class.

magnatelee replied:
Also, could we make the constructor take an optional stream argument? Since this PR is going to change the join interface anyway, this seems to be the right chance to expose asynchrony for hash join. @jrhemstad What do you think?

jrhemstad replied:
The OO `hash_join` interface is still a public interface and we don't want to expose streams in public interfaces yet.

zanmato1984 replied:
The static factory method and the subsequent virtual `*_join` methods are consequences of that I was trying to hide the actual `multimap_type` type and member from the public header file.

But if that is not a concern, I'll save the aforementioned static/virtual methods. Any suggestions?

jrhemstad replied:
Ah... I see. That makes more sense now. I agree that it seems worthwhile to avoid exposing the `multimap` type. My preference would be to use PImpl to accomplish the same goal. That way we can avoid the awkwardness of the static factory.

zanmato1984 replied:
Yeah PImpl sounds better. Thanks for reminding, will do.

zanmato1984 replied:
Addressed.

jrhemstad commented on cpp/include/cudf/join.hpp:
```diff
@@ -367,5 +367,131 @@ std::unique_ptr<cudf::table> cross_join(
+   * specified by `build_on` and `probe_on`. The resulting table will be joined columns of
+   * `probe(including common columns)+build(excluding common columns)`.
+   */
+  virtual std::unique_ptr<cudf::table> full_join(
```
Do these need to be virtual? I don't expect we're going to have different derived implementations of `hash_join`.

zanmato1984 replied:
Ditto.

zanmato1984 replied:
Addressed.

jrhemstad commented on cpp/tests/join/join_tests.cpp:
```diff
@@ -56,6 +127,13 @@ TEST_F(JoinTest, InvalidCommonColumnIndices)
 
   EXPECT_THROW(cudf::inner_join(t0, t1, {0, 1}, {0, 1}, {{0, 1}, {1, 0}}), cudf::logic_error);
+
+  EXPECT_THROW((build_once_probe_multiple_parallel<join_kind::INNER, false>(
```
The OO interface should have it's own tests and not be part of the existing tests. The existing tests will already exercise the OO interface to make sure it preserves existing functionality. Any new functionality (such as probing with multiple tables) should be new tests.

zanmato1984 replied:
Yes, will do.

zanmato1984 replied:
I'm thinking about adding new tests of probing multiple/single table in parallel, and specifying probe output side. These are the new functionalities I can think of. Any other things that I may have missed?

And one more question, should I add tests for all the corner cases that existing tests have already covered? Such as empty table on either/both sides, null equality, etc.

jrhemstad replied:
> should I add tests for all the corner cases that existing tests have already covered? Such as empty table on either/both sides, null equality, etc.

Since you've refactored the existing public Join APIs to use the OO interface, I think the existing tests are sufficient to exercise this behavior.

I think the only new tests needed are for exercising the new behavior, i.e., probing a single table with multiple other tables. 

I don't think we need to provide tests for probing in parallel. libcudf makes very weak guarantees about thread safety. So the multi-probe tests should just sequentially probe the `hash_join` object with multiple probe tables.

zanmato1984 replied:
Yes, will do.

zanmato1984 replied:
Addressed.

zanmato1984 commented:
Thanks for reviewing, @jrhemstad . I have some more comments that needs your kind input. PTAL.

jrhemstad commented on cpp/include/cudf/join.hpp:
```diff
@@ -367,5 +367,138 @@ std::unique_ptr<cudf::table> cross_join(
+  hash_join& operator=(hash_join&&) = delete;
+
+  /**
+   * @brief Construct a hash join object and build the internal hash table used for subsequent
```
When the hash table is built is an implementation detail. E.g., it could be built lazily on the first `*_join` invocation. 

We should also make sure to indicate that this object does not own the `build` table and its lifetime needs to be maintained by the user. 
```suggestion
   * @brief Construct a hash join object for subsequent probe calls. 
   *
   * @note The `hash_join` object must not outlive the table viewed by `build`, else behavior is undefined.
```

jrhemstad commented on cpp/include/cudf/join.hpp:
```diff
@@ -367,5 +367,138 @@ std::unique_ptr<cudf::table> cross_join(
+  /**
+   * @brief Performs an inner join by probing in the internal hash table.
+   *
+   * Given that it is sometimes desired to choose the small table to be the `build` side for an
+   * inner join (which is automatically done in cudf::innner_join()), it is handy to be able to
+   * specify which side in the joined table the `probe` portion is on. That is, one can check
+   * whether to use `left` or `right` table to be the build side by comparing their row counts ahead
+   * of `hash_join` object, but still preserve the logical order of the `left`/`right` in the joined
+   * table. For instance, if the `left` table is the small one (and thus the `build` side),
+   * specifying the `probe_output_side` to `RIGHT` will output the `right` table (probe side) on the
+   * right side in the joined table.
```
So this parameter is kind of awkward. I think we could eliminate it by making these member functions return a `pair<unique_ptr<table>, unique_ptr<table>>` of the columns from the `build` and `probe` tables respectively. Then at the call site they can be swapped if necessary.

zanmato1984 replied:
I think caller can always adjust the order of `build`/`probe` columns, and it is not much harder than adjusting the order of a packed `build`/`probe` table pair.

If the concern is to keep the API simple, maybe we can just remove this parameter?

jrhemstad replied:
Yeah, I think the `probe_output_side` parameter is confusing. 

So you're suggesting return a single `table` and just switch the order of the columns in that table? I think that can be tricky, especially due to the `columns_in_column` stuff.

zanmato1984 replied:
I mean returning a single table and the columns order being **strictly** of `probe` on the left and `build` on the right. Users can adjust columns order within the returned table.

zanmato1984 replied:
Additionally, I think returning `build`/`probe` table pair would be overkill, since getting a single result table from a `*_join` call seems like a more common case.

jrhemstad replied:
> Users can afterwards adjust columns order within the returned table on their own demands.

But I think `columns_in_common` make it tricky to know what columns correspond to `build` vs `probe`. Maybe I'm mistaken. Could you try implementing the existing `cudf::inner_join` w/o the `probe_output_side` parameter and manually rearranging the columns to show what it would take?

zanmato1984 replied:
I think `columns_in_common` can still go as described in the old `cudf::inner/left/full_join` functions, i.e. `build` always being right and `probe` being left and `columns_in_common` being gathered from `probe` (left).

The rearranging code would be like:
`
table_view t0(a0, b0, c0);
hash_join hj(t0, {0, 1});
table_view t1(a1, b1, c1);
result = hj.inner_join(t1, {0, 1}, {{0, 0}, {1, 1}});
// Refer to the description of `columns_in_common`, user will know result is (a1, b1, c1, c0).
// Switching left/right into a new table view will be like:
table_view rearranged(result[0], result[1], result[3], result[2]);
`
Does this answer the question?

jrhemstad replied:
We need to implement that logic when implementing the existing `cudf::inner_join` in terms of the new `hash_join` OO implementaiton.

zanmato1984 replied:
> We need to implement that logic when implementing the existing `cudf::inner_join` in terms of the new `hash_join` OO implementation.

Sorry, I don't think I'm following. Could you explain more?

zanmato1984 replied:
I think I got ya. Let me think a moment.

zanmato1984 replied:
You were right. It will be non-trivial to implement the existing `cudf::inner_join` using `hash_join` and rearranging the resulting table.

But I'm thinking returning `build`/`probe` table pair is equally non-trivial in terms of rearranging, since the "common" columns have to be in either side (but not both, otherwise they will be materialized twice), the trouble of `columns_in_common` still exists in the same way as in the single resulting table.

Moreover, returning `build`/`probe` table pair makes `hash_join::inner_join` not quite aligned with the other two `hash_join::left_join` and `hash_join::full_join` members in terms of signature, and users pay more attention of combining them despite they want to rearrange them or not.

What do you think?

jrhemstad replied:
> since the "common" columns have to be in either side (but not both, otherwise they will be materialized twice), the trouble of columns_in_common still exists in the same way as in the single resulting table.

Won't this be taken care of automatically be swapping the values in `columns_in_common`? 

> Moreover, returning build/probe table pair makes hash_join::inner_join not quite aligned with the other two hash_join::left_join and hash_join::full_join members in terms of signature,

I'm not too concerned about this. This `hash_join` object is already intended as more of an "expert-mode" API anyways.

zanmato1984 replied:
Oh, let me explain more about the "non-trivial" part.

Image that we end up with the following `hash_join::inner_join` API (w/o the `probe_output_side` parameter):

`std::pair<table /*probe portion*/, table /*build portion*/> hash_join::inner_join(probe, probe_on, columns_in_common, mr);`

1. The order of probe/build portions needs to be fixed as there is no flag in the parameter list to specify it. We can inform users this spec through proper comments. This wouldn't be an issue.
2. Due to the existence of `columns_in_common`, the "common" columns need to go into either probe or build portion. We can too fix this spec through comments and inform users, say we go with "common columns will be output in the probe portion".

Now consider the old `cudf::inner_join` API, under the circumstances that `left` being smaller:
```
hash_join(left, left_on);
swap(columns_in_common);
[probe, build] = hash_join.inner_join(right, right_on, columns_in_common);
// Now do rearrangement to preserve the logical left/right order.
probe_common_indices = ...;
probe_noncommon_indices = ...;
common_columns = probe.select(probe_common_indices);
right_noncommon_columns = probe.select(probe_noncommon_indices);
left_noncommon_columns = build;
left_columns = ...; // Zip `left_noncommon_columns` with `common_columns` to restore the relative orders of the columns in the original `left` table.
return table(combine(left_columns, right_noncommon_columns));
```

Likewise, for the direct `hash_join::inner_join` users, once they want to explicitly swap the build/probe tables, they'll have to do the last 8 lines of the above code block. IMO this would be cumbersome even for "expert" users.

Please help to see if my explanation make sense.

zanmato1984 replied:
Any comment?

zanmato1984 replied:
@jrhemstad Hi Jake, could you please take a look the comments above? Thanks.

jrhemstad replied:
I had to walk through a concrete example before I could see what the issue is. Here's the example I went through:

Imagine two tables:
```
table p : columns {a, b}
table q : columns {d, a}
```

We want to join "on" the `a` column from each table. Furthermore, we want to follow Pandas' semantics and only materialize one copy of `a` since it has the same name. 

So calling `cudf::inner_join` would look like:
```
cudf::inner_join(p, q, left_on{0}, right_on{1}, common{ {0,1} })
```

Since `p` is the "left" table, we want to keep the copy of `a` with `p`. 

Now we have two situations to consider.

First, consider `p.num_rows() > q.num_rows()`. 

`q` is the smaller table, so we build on `q`. This is swapping the original (p,q) order, so we need to take care of that swapping.

```
[q', p'] = hash_join obj(q, on{1}).inner_join(p, on{0}, common{1,0});
```

The tables `q'` and `p'` hold the join results from `q` and `p` respectively. 

**So here is the problem:** If we say "common columns always go with the build table", then `q'` will contain the `a` column, but we wanted it in `p'`! Instead, if we say "common columns always go with the probe table", then it will break in the situation where `p.num_rows() < q.num_rows()`. 

So sadly we do need some kind of parameter that controls where the "common" columns go, but this has become very convoluted. 

My preference would still to make `inner_join` return a `pair<table,table>` for build,probe respectively, but change the `probe_output_side` argument to instead control whether common columns are placed in `build` or `probe`.

jrhemstad replied:
I'll open an issue to get rid of `columns_in_common` all together by instead having APIs to simply return the indices into the left and right table of the matching rows.

zanmato1984 replied:
So at least we both now agree that an extra parameter would be necessary to ease the rearrangement of common columns in the output table(s): either `probe_output_side` or something like `common_columns_go_with` as you proposed.

I can do the `common_columns_go_with` way. But I'm not sure about your latest comment about getting rid of `columns_in_common`. I think changing this PR to the `common_columns_go_with` way could be done in parallel with getting rid of `columns_in_common`. If so, I'll go on changing the code. Or should I wait for your new APIs and leverage them after they are done?

Please advice. Thanks.

jrhemstad replied:
> I think changing this PR to the common_columns_go_with way could be done in parallel with getting rid of columns_in_common. If so, I'll go on changing the code. Or should I wait for your new APIs and leverage them after they are done?

Getting rid of `columns_in_common` will be future work and should not block this PR. Proceed with making the minimal changes to this PR and we'll get it merged.

zanmato1984 replied:
Addressed.

jrhemstad commented on cpp/include/cudf/join.hpp:
```diff
@@ -367,5 +367,138 @@ std::unique_ptr<cudf::table> cross_join(
+   *
+   * @param probe The probe table, from which the tuples are probed.
+   * @param probe_on The column indices from `probe` to join on.
+   * @param columns_in_common is a vector of pairs of column indices into
+   * `build` and `probe`, respectively, that are "in common". For "common"
+   * columns, only a single output column will be produced, which is gathered
+   * from `probe_on` columns or `build_on` columns if `probe_output_side` is LEFT or RIGHT.
+   * Else, for every column in `probe_on` and `build_on`,
+   * an output column will be produced. For each of these pairs (P, B), P
+   * should exist in `probe_on` and B should exist in `build_on`.
```
I really dislike this whole `columns_in_common` business we had in the old API. The whole point of it is to avoid unnecessarily materializing columns that Pandas wouldn't materialize because the columns have the same name. Don't worry about it in this PR, but I think we need to find a better way to do this. Perhaps by having APIs that just return the indices of matching rows and then Python can handle gathering from the appropriate columns.

zanmato1984 commented:
Hi @jrhemstad, I found some minor bugs when dealing with `columns_in_common` for inner joins when discussing with you. I've fixed them and added some covering tests in the latest commit.

These are aside of our discussion so this is just FYI. Sorry for the inconvenience for reviewing.

jrhemstad commented on cpp/include/cudf/join.hpp:
```diff
@@ -367,5 +367,140 @@ std::unique_ptr<cudf::table> cross_join(
+  hash_join(cudf::table_view const& build, std::vector<size_type> const& build_on);
+
+  /**
+   * @brief Side of the probe table in the joined table. Only applicable for inner join.
```
```suggestion
   * @brief Controls the location in the output of the columns from the `probe` table. 
```

jrhemstad commented on cpp/include/cudf/join.hpp:
```diff
@@ -367,5 +367,140 @@ std::unique_ptr<cudf::table> cross_join(
+  /**
+   * @brief Side of the probe table in the joined table. Only applicable for inner join.
+   */
+  enum class probe_output_side { LEFT, RIGHT };
```
If we do end up keeping this, I think we can improve the docs to make it a bit easier to understand. 
```suggestion
  enum class probe_output_side { 
     LEFT, ///< Columns from the probe table appear as the leftmost in the output table
     RIGHT ///< Columns from the probe table appear as the rightmost in the output table 
  };
```

zanmato1984 commented:
@jrhemstad Hi Jake, could you please take a look at the comments in our discussion? Thanks.

jrhemstad commented:
> @jrhemstad Hi Jake, could you please take a look at the comments in our discussion? Thanks.

Thanks for your patience. I'll review today.

zanmato1984 commented:
@jrhemstad Hi Jake, I've addressed your last comment. PTAL, thanks.

rgsl888prabhu commented on cpp/tests/join/join_tests.cpp:
```diff
@@ -525,6 +525,96 @@ TEST_F(JoinTest, InnerJoinNoNulls)
 
+TEST_F(JoinTest, InnerJoinNonAlignedCommon)
+{
+  column_wrapper<int32_t> col0_0{{3, 1, 2, 0, 2}};
```
Just to reduce number of lines, nothing else
```suggestion
  cols0.push_back(column_wrapper<int32_t> ({3, 1, 2, 0, 2}).release());
```

zanmato1984 replied:
Addressed.

zanmato1984 commented:
@rgsl888prabhu Hi Ram, thanks for reviewing. Comments addressed. PTAL, thanks.

zanmato1984 commented:
How do I trigger CI to proceed merging? I saw other PRs triggered CI by updating CHANGELOG, however I didn't receive instructions like "Please update the changelog in order to start CI tests" from GPUtester.

Shall I go on anyway?

rgsl888prabhu commented:
I needs to have a update in CHANGELOG, not sure why it is not triggering any issues.

rgsl888prabhu commented:
rerun tests

zanmato1984 commented:
rerun tests

raydouglass commented:
ok to test

kkraus14 commented:
rerun tests

kkraus14 commented:
On the Python side there's some tests that use `NaN` data that are throwing:
```
vector::_M_range_check: __n (which is 1) >= this->size() (which is 1)
```

There's also some weird `ColumnAccessor` failures that I'm not sure about. cc @shwina

zanmato1984 commented:
> On the Python side there's some tests that use `NaN` data that are throwing:
> 
> ```
> vector::_M_range_check: __n (which is 1) >= this->size() (which is 1)
> ```
> 
> There's also some weird `ColumnAccessor` failures that I'm not sure about. cc @shwina

The failures may have something to do with "empty" data set. Looking.

zanmato1984 commented:
@kkraus14 Hi Keith, CI fixed, could you please help to proceed? Thanks.

kkraus14 commented:
> @kkraus14 Hi Keith, CI fixed, could you please help to proceed? Thanks.

Thanks for tackling this! @jrhemstad can you review and merge if it looks good to you?

jrhemstad commented:
@zanmato1984 is there a C++ unit test we could add that would have caught this error?

zanmato1984 commented:
> @zanmato1984 is there a C++ unit test we could add that would have caught this error?

Yes, I did create a C++ case in my local to trace this error. It's basically a translation of this python case: https://github.com/rapidsai/cudf/blob/afb3928f64a4f7d4f8a21ed12ef3a17353f42f54/python/cudf/cudf/tests/test_joining.py#L802

I thought as python case will cover it anyhow so I didn't add the C++ case.

Do you think we still need that C++ case?

kkraus14 commented:
> Do you think we still need that C++ case?

Generally we want C++ coverage of all things if possible as it removes the Python layer from the equation and allows for easier debugging / triage.

zanmato1984 commented:
C++ tests added, @jrhemstad PTAL thanks.

kkraus14 commented:
ok to test

zanmato1984 commented:
rerun tests

kkraus14 commented:
@zanmato1984 don't worry about this failure, just waiting for an approval from @jrhemstad and I'll admin merge.

Thank you for all of your hard work on this! It's greatly appreciated!

jrhemstad commented:
> @zanmato1984 don't worry about this failure, just waiting for an approval from @jrhemstad and I'll admin merge.
> 
> Thank you for all of your hard work on this! It's greatly appreciated!

I approved a while back. It's good to go.
